bus-mapping/src/circuit_input_builder.rs

```diff
@@ -761,93 +787,172 @@ impl<'a> CircuitInputStateRef<'a> {
+                    Some(OpEnum::Account(operation.op().reverse()))
+                } else {
+                    None
+                }
+            }
+            OperationRef(Target::AccountDestructed, idx) => {
+                let operation = &self.block.container.account_destructed[*idx];
+                if operation.rw().is_write() && operation.reversible() {
+                    Some(OpEnum::AccountDestructed(operation.op().reverse()))
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
+
+    /// Apply reverted op to state and push to container.
+    fn apply_reverted_op(
+        op: OpEnum,
+        block: &mut Block,
+        block_ctx: &mut BlockContext,
+        sdb: &mut StateDB,
+    ) {
+        match op {
+            OpEnum::Storage(op) => {
+                let (_, account) = sdb.get_storage_mut(&op.address, &op.key);
+                *account = op.value;
+                block
+                    .container
+                    .insert(Operation::new(block_ctx.rwc.inc_pre(), RW::WRITE, op));
+            }
+            OpEnum::TxAccessListAccount(op) => {
+                if !op.value {
+                    sdb.remove_account_from_access_list(&op.address);
+                }
+                block
+                    .container
+                    .insert(Operation::new(block_ctx.rwc.inc_pre(), RW::WRITE, op));
+            }
+            OpEnum::TxAccessListAccountStorage(op) => {
+                if !op.value {
+                    sdb.remove_account_storage_from_access_list(&(op.address, op.key));
+                }
+                block
+                    .container
+                    .insert(Operation::new(block_ctx.rwc.inc_pre(), RW::WRITE, op));
+            }
+            OpEnum::Account(op) => {
+                let (_, account) = sdb.get_account_mut(&op.address);
+                match op.field {
+                    AccountField::Nonce => account.nonce = op.value,
+                    AccountField::Balance => account.balance = op.value,
+                    AccountField::CodeHash => {
+                        account.code_hash = op.value.to_be_bytes().into();
                     }
                 }
+                block
+                    .container
+                    .insert(Operation::new(block_ctx.rwc.inc_pre(), RW::WRITE, op));
+            }
+            OpEnum::TxRefund(_) => unimplemented!(),
+            OpEnum::AccountDestructed(_) => unimplemented!(),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Handle a reversion caused by call_idx
+    fn handle_reversion(&mut self, call_idx: usize) {
+        let mut step_index = self.tx.steps.len() - 1;
+
+        loop {
+            let step = &self.tx.steps[step_index];
+
+            // Skip failure callees
+            if step.call_index != call_idx && !self.tx.calls[step.call_index].is_success {
```

Comment:
I didn't think about doing it this way!  It's much better than my approach: more clear and you jump over already reverted steps :D